FPU operator

wardvermeulen opened the pull request:
This PR introduces a new operator `FPUOp` in an attempt to make GemmKernels.jl function on any CUDA-enabled GPU, i.e. on older GPUs that do not support WMMA. 

The `FPUOp{M, N, K, DT, CT}` operator type has five parameters. (M, N, K) denotes the shape of the operator at the warp level. The base shape of this operator is (4, 8, 1). Every other shape must consist of multiples of these base sizes. For example (8, 8, 1) or (16, 8, 4) are allowed, but not (2, 16, 1). DT is the data type of the C and D matrices of the GEMM, and CT is the compute type of the GEMM, i.e. the data type of matrices A and B. 

## Evaluation

An operator shape of (8, 8, 1) was used each time for the GemmKernels.jl benchmarks. Other parameters such as block sizes, grid sizes, etc. were tuned manually until a decent result was found. Perhaps faster configurations exist.
### GeForce RTX 2080 Ti benchmark

![gemm](https://github.com/JuliaGPU/GemmKernels.jl/assets/37931310/5798911f-3602-48c2-bef5-c29e02539109)

### Tesla V100 benchmark

![gemmV100](https://github.com/JuliaGPU/GemmKernels.jl/assets/37931310/f1b913eb-5f1c-40f9-a5ba-2d75aae9ea14)

## Limitations

- Mixed typing is supported and functions correctly. `Float16 x Float16 + Float32` is possible, but I have not found a configuration yet where that is faster than `Float32 x Float32 + Float32`. 
- `Float16 x Float16 + Float16` is really imprecise, and fails tests, which is why it is not included in the `matmul.jl` test file.
- apply_iterate of #99 is more prevalent when using `FPUOp` because the operator size is much smaller than for WMMA. Each LocalArray can hold at max 16 elements. The block shapes should therefore adhere to the following rule: 

`(blockshape.M / operatorshape.M) * (blockshape.N / operatorshape.N) <= 16`

thomasfaingnaert commented on benchmarks/fpu/plot.jl:
```diff
@@ -0,0 +1,40 @@
+    savefig(p, "gemm.png")
+end
+
+isinteractive() || main(ARGS[1])
```
What is the purpose of `ARGS[1]` here?

thomasfaingnaert reviewed (commented):
Looks good! Just added some minor comments.

thomasfaingnaert commented:
>     `Float16 x Float16 + Float32` is possible, but I have not found a configuration yet where that is faster than `Float32 x Float32 + Float32`.

That's strange. Do you take into account that if the A and B matrix are Float16, they take up only half the space in global and shared memory, and hence that you can double the tile sizes when comparing Float16 and Float32?

wardvermeulen commented:
>  That's strange. Do you take into account that if the A and B matrix are Float16, they take up only half the space in global and shared memory, and hence that you can double the tile sizes when comparing Float16 and Float32?

I should experiment further with this, but should it not it not already be faster without doubling the tile sizes since double the amount of elements can be SIMD loaded?

wardvermeulen commented on benchmarks/fpu/benchmark.jl:
```diff
@@ -0,0 +1,113 @@
+using JSON
+
+PATH_TO_NCU = haskey(ENV, "PATH_TO_NCU") ? ENV["PATH_TO_NCU"] : "/usr/local/NVIDIA-Nsight-Compute-2023.1/ncu"
+PATH_TO_JULIA = haskey(ENV, "PATH_TO_JULIA") ? ENV["PATH_TO_JULIA"] : "/home/wjvermeu/julia-versions/julia-1.8.3/bin/julia"
+PATH_TO_CUTLASS = haskey(ENV, "PATH_TO_CUTLASS") ? ENV["PATH_TO_CUTLASS"] : "/home/wjvermeu/cutlass/build/tools/profiler/cutlass_profiler"
```
Is it sufficient to change it to: `PATH_TO_NCU = ENV["PATH_TO_NCU"]` since this will throw a `KeyError` if the environment variable is not defined?

maleadt replied:
Or just do something like `get(ENV, "PATH_TO_NCU", "ncu")` which would then assume `ncu` to be on PATH unless you specify the env var.

wardvermeulen commented on src/operator.jl:
```diff
@@ -13,6 +16,123 @@ for f in (:fragtype_a, :fragtype_b, :fragtype_accum, :load_a, :load_b, :load_c,
+    @unroll for k = 1 : K
+        @unroll for m = 1 : M ÷ 4
+            @unroll for n = 1 : N ÷ 8 
+                if (DT == CT)
```
I forgot `fma` actually already casts the arguments to the needed type and is much faster, so this will not be necessary after all.

wardvermeulen commented on src/operator.jl:
```diff
@@ -13,6 +16,123 @@ for f in (:fragtype_a, :fragtype_b, :fragtype_accum, :load_a, :load_b, :load_c,
+                    # Use fma
+                    @inbounds c_frag = setindex(
+                        c_frag,
+                        DT(fma(a_frag[m + (M ÷ 4) * (k - 1)], b_frag[k + K * (n - 1)], c_frag[m + (M ÷ 4) * (n - 1)])),
```
```julia
    a_frag = LocalArray{Tuple{M ÷ 4, K}, CT}(a_frag)

    @unroll for k = 1 : K
        @unroll for m = 1 : M ÷ 4
            @unroll for n = 1 : N ÷ 8 
                @inbounds c_frag = setindex(
                    c_frag,
                    fma(
                        a_frag[m, k], 
                        b_frag[k + K * (n - 1)], 
                        c_frag[m + (M ÷ 4) * (n - 1)]
                    ),
                    m + (M ÷ 4) * (n - 1)
                )
            end
        end
    end

    return NTuple{M * N ÷ 32}(c_frag)
```
This, for example, requires 198 registers instead of 168 according to Nsight Compute, increasing the execution time by approx. 1000x.

maleadt replied:
Can you diff the generated code? I had a quick look, and ND indexing of LocalArray seems to generate the same IR. Maybe this is just a missing `@inline`?

wardvermeulen replied:
`@inline` seems to have fixed it. Thank you.

maleadt replied:
Great!

wardvermeulen commented:
I fixed the 2D indexing of the fragments and changed the accesses to be in column-major order, just to be concise, there is no performance difference. I also added support for other FPU operators with different arithmetic, and included the example for tropical arithmetic as previously discussed in #51.

thomasfaingnaert commented on src/operator.jl:
```diff
@@ -13,6 +16,131 @@ for f in (:fragtype_a, :fragtype_b, :fragtype_accum, :load_a, :load_b, :load_c,
+for (operator_type) in [FPUOp, TropicalFPUOp]
+    @eval begin
+        @inline function mma(::Type{$operator_type{M, N, K, DT, CT}}, a_frag, b_frag, c_frag) where {M, N, K, DT, CT}
+            a_frag = LocalArray{Tuple{M ÷ 4, K}, CT}(a_frag)
+            b_frag = LocalArray{Tuple{K, N ÷ 8}, CT}(b_frag)
+            c_frag = LocalArray{Tuple{M ÷ 4, N ÷ 8}, DT}(c_frag)
```
Is the casting necessary here? That is to say, can't you just return 2D arrays in `load_a`, `load_b`, and `load_c` in the first place?

wardvermeulen replied:
No, that was actually one of the major pain points in this implementation. Returning a `LocalArray` results in a segmentation fault and crashes the Julia REPL. You can verify this by just changing the return type for the A fragment, returning frag in `load_a` and commenting out the first line of this snippet. Casting to a `Tuple` was the workaround I managed to find to mitigate this issue.

wardvermeulen commented on src/operator.jl:
```diff
@@ -13,6 +16,131 @@ for f in (:fragtype_a, :fragtype_b, :fragtype_accum, :load_a, :load_b, :load_c,
+
+for (operator_type) in [FPUOp, TropicalFPUOp]
+    @eval begin
+        @inline function mma(::Type{$operator_type{M, N, K, DT, CT}}, a_frag, b_frag, c_frag) where {M, N, K, DT, CT}
```
Okay, I will change this.
